content/post/2021-07-12-code-detective/index.Rmd

```diff
@@ -0,0 +1,267 @@
+
+Easy case: there's a message on screen telling you where an error occurred, or you know what function you want to amend.
+
+Alternatively,
+
+* You can put the error / warning in a search engine.
+* If there's an unclear error you can try to see the traceback i.e. what functions were called leading to that error. In her talk ["Object of type 'closure' is not subsettable"](https://github.com/jennybc/debugging#readme), Jenny Bryan explains very well what a traceback is. In my `.Rprofile` I have
+
+```r
+options(
+error = rlang::entrace, 
+rlang_backtrace_on_error = "branch")
+```
+
+thanks to a [tweet by Noam Ross](https://twitter.com/noamross/status/1202269314029621251) reporting a tip by Jenny Bryan. "It gives trimmed tracebacks when using pipes."
+
+* If there's no error but say a warning you could try to [convert the warning to an error](https://adv-r.hadley.nz/debugging.html#non-error-failures).
+
+### Explore from that starting point
+
+* You can use "grepping" as said by Patricia Aas: look for the occurrences of a function or variable name in a local folder, or via [GitHub (advanced) search](https://docs.github.com/en/github/searching-for-information-on-github/getting-started-with-searching-on-github/about-searching-on-github). You can limit the hits to some types of files e.g. R scripts in `R/`.
+* In your IDE e.g. RStudio there might be a way to go directly to the _definition_ of function. With the [lookup package](https://github.com/jimhester/lookup) you can also easily look up the source of functions [locally and on GitHub](https://blog.r-hub.io/2019/05/14/read-the-source/).
+
+### How to read code: space... and time
+
+* Hopefully the code makes sense on its own. If not, fear not, the next item right here, and the other sections of this post, as well as patience, will help.
+* Sometimes [using git blame or looking at the git history](https://docs.github.com/en/github/managing-files-in-a-repository/managing-files-on-github/tracking-changes-in-a-file) might help understand the context of some aspects of the code, if there's no code comment referring an issue. Do not actually _blame_ people, though. To make your own git history more informative for such later code archaeology, use branches and squash and merge.
+
+## Build your mental model of the code
+
+That's what [Patricia Aas calls "mental machine"](https://www.allthingsgit.com/episodes/learning_a_new_codebase_with_patricia_aas.html).
+You might want to draw some sort of diagram by hand (or [programmatically](https://blog.r-hub.io/2019/12/12/internal-functions/#explore-internal-functions-within-a-package)).
+Patricia Aas remarks that such diagrams might even be contributed to the codebase as developer documentation.
+
+## Browse code by others
+
+The life-hack below by Julia Silge for fixing Travis CI builds, looking at other people's configuration files, applies to other CI services (good as Travis is e.g. no longer recommended by rOpenSci) and to any other code endeavour.
```

Comment:
```suggestion
The life-hack below by Julia Silge for fixing Travis CI builds, looking at other people's configuration files, applies to any code endeavour.
```
I'd shorten it to "applies to any code endeavour" since that is the main point of this sentence. If you want to add a warning about Travis, I'd put that in parenthesis.